server/src/main/java/io/deephaven/server/notebook/FilesystemStorageServiceGrpcImpl.java

```diff
@@ -0,0 +1,251 @@
+import io.deephaven.configuration.Configuration;
+import io.deephaven.extensions.barrage.util.GrpcUtil;
+import io.deephaven.internal.log.LoggerFactory;
+import io.deephaven.io.logger.Logger;
+import io.deephaven.proto.backplane.grpc.CreateDirectoryRequest;
+import io.deephaven.proto.backplane.grpc.CreateDirectoryResponse;
+import io.deephaven.proto.backplane.grpc.DeleteItemRequest;
+import io.deephaven.proto.backplane.grpc.DeleteItemResponse;
+import io.deephaven.proto.backplane.grpc.FetchFileRequest;
+import io.deephaven.proto.backplane.grpc.FetchFileResponse;
+import io.deephaven.proto.backplane.grpc.ItemInfo;
+import io.deephaven.proto.backplane.grpc.ItemType;
+import io.deephaven.proto.backplane.grpc.ListItemsRequest;
+import io.deephaven.proto.backplane.grpc.ListItemsResponse;
+import io.deephaven.proto.backplane.grpc.MoveItemRequest;
+import io.deephaven.proto.backplane.grpc.MoveItemResponse;
+import io.deephaven.proto.backplane.grpc.SaveFileRequest;
+import io.deephaven.proto.backplane.grpc.SaveFileResponse;
+import io.deephaven.proto.backplane.grpc.StorageServiceGrpc;
+import io.deephaven.server.session.SessionService;
+import io.grpc.stub.StreamObserver;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.DirectoryNotEmptyException;
+import java.nio.file.DirectoryStream;
+import java.nio.file.FileAlreadyExistsException;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.NoSuchFileException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardOpenOption;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.regex.PatternSyntaxException;
+
+import static com.google.common.io.Files.asByteSource;
+
+/**
+ * Default implementation for the StorageService gRPC service, which will use the filesystem to store files on behalf of
+ * authenticated clients.
+ *
+ * Current implementation only checks if a user is logged in, and doesn't provide finer grained access controls to
+ * files.
+ */
+@Singleton
+public class FilesystemStorageServiceGrpcImpl extends StorageServiceGrpc.StorageServiceImplBase {
+    private static final Logger log = LoggerFactory.getLogger(FilesystemStorageServiceGrpcImpl.class);
+
+    private static final String STORAGE_PATH =
+            Configuration.getInstance().getStringWithDefault("storage.path", "<workspace>/storage")
+                    .replace("<workspace>", Configuration.getInstance().getWorkspacePath());
+
+    /**
+     * Non-cryptographic hash, not resistant to adversarial collisions, but should suffice for quickly checking for
+     * edits to files. By design, hashes might not be stable from one startup to the next, but will at least efficiently
+     * guard against client's needing to load the same file again rapidly.
+     */
+    private static final HashFunction HASH_FUNCTION = Hashing.goodFastHash(128);
```

Comment:
Discussed on phone, I'd be in favor of `com.google.common.hash.Hashing#murmur3_128(int)` with a fixed seed. I think we should still have the comment that "the client should not *rely* on this implementation and it _may_ change from run to run"